src/server/pkg/storage/chunk/reader.go

```diff
@@ -33,6 +33,11 @@ func newReader(ctx context.Context, objC obj.Client, prefix string, dataRefs []*
 	}
 }
 
+func (r *Reader) RangeSet(dataRefs []*DataRef) {
+	r.dataRefs = dataRefs
+	r.r = bytes.NewReader([]byte{})
+}
```

Comment:
Why would you want to call this method instead of just constructing a new Reader? It seems like it basically creates a new `Reader` from scratch, except that it doesn't clear `buf` is that significant? Would the semantics of this function change if you added `buf = &bytes.Buffer{}`?

Also the name on this function is really weird. I spent a long time trying to figure out what `RangeSet` was until I realized that "set" is a very here not a noun. It seems very non-standard and confusing the have a setter function named `NounSet` rather than `SetNoun`.

Reply:
The reader has the current chunk cached (buf). It will reuse that chunk if it makes sense (curr.Hash == dataRefs[0].Hash). Therefore, if the client is receiving a stream of dataRefs it can make RangeSet calls as it gets them (of course reading to EOF between) to get efficient reads (as I described in the description). This is the exact access pattern for the fileset reader when reading the index.

This naming convention is what kubernetes uses for kubectl (`kubectl config set`) for example. It makes sense to me because it is easier to group operations associated with a concept, but I agree that it reads kind of weird. Again, similar to the chunk storage PR, no strong opinion, but if you are strongly opposed I am not against changing the convention.

Reply:
I see, that makes more sense. The interface still seems a bit wonky to me, what would you think of just having `NewReader` take an optional `*Reader` parameter called `oldReader` that it could pull the buffer from if it was for the same chunk. I think that would make it a bit clearer that the intended use for this is the make starting a new read with a common backing chunk more efficient.

Reply:
I considered that model, but I like what I have now better. What I have now is pretty analogous to the way reading a tar file works (next then read until EOF). I just think that it is a cleaner and easier to use abstraction, and the semantics seem pretty sensible to me. I guess the only thing that could seem wonky is that the set operation takes arbitrary data refs, but that just seems like a plus in terms of usability.

Reply:
I don't think it's very analogous to the way Tar readers work. With Tar readers you specify up front everything you want to read and then `Next` is just used to advance through it. Here you're reusing a reader struct for multiple reads and the relationship between those reads isn't immediately clear to the caller, they'd need to realize that `DataRef`s might reference the same underlying data and thus there's a chance for an optimization here. But if someone just starts using this interface they likely won't realize that, and they'll probably just call `NewReader` every time, since semantically that does what they want, and just ignore the `RangeSet` method. If it's an argument to `NewReader` then you can't help but understand it, since you have to call that method.

Reply:
If your concern is about a client of this API using it in a way that does not take advantage of this optimization, then we can simply document that you should setup one reader for a stream of data rather than multiple. Having to setup new readers for every new set of data refs and seeding the prior just seems messy and adds no value if we communicate how to use the API correctly. The optimization is not something that the client should have to think about. As far as being analogous to tar, it is definitely similar in respect to the access pattern regardless of whether all the data is specified up front. You are essentially using `RangeSet` to seek through a logical byte stream.

Reply:
I generally disagree with the argument that a confusing API + documentation is a substitute for a non-confusing interface.

If you're thinking of this as similar to the tar interface with its `Next` method how would you feel about calling this method `NextRange`? And leaving the functionality as is. That seems easier to understand to me, although I still think it's a kind of confusing interface.

Also going back to your previous comment on organizing of names, I don't think we should start having interfaces be `NounVerb` we have too many interfaces that are already `VerbNoun` for this switch to be consistent, and I think that reasoning makes a lot more sense in CLIs where you want to use tab completion to find all the methods that act on an certain type of object.

Reply:
I am not making the argument that a confusing API + docs is a substitute for a non-confusing API because I do not think this is a confusing API. I like that clients do not have to think about optimizations and I believe the optimization will naturally match with their access pattern. If they are doing a bunch of random accesses, then they probably won't be keeping track of readers that may have chunks cached and if they are doing sequential accesses then the optimization just magically happens.

`NextRange` is a much better name, I do not know why I did not think of that lol. Also, I am fine with sticking to `VerbNoun` if it reads easier and is consistent. The more I think about it, it probably is not that much value to follow the `NounVerb` convention for function names because there are not many times when having the common prefix is useful.

Reply:
I think you still have to think about optimizations here, because you have to know to call `NextRange` instead of just creating a `NewReader`, but I think that's tolerable and having to think about optimizations is better than having them happen in magical and potentially more confusing ways. Let's just got with `NextRange` as a name and call it a day.